#' Find Differentially Methylated Regions across whole genome
#'
#' A function to define DMR for an entire genome
#'
#' @param tumor_table A matrix of beta-values (percentage) from tumor samples.
#' @param control_table A matrix of beta-values (percentage) from
#' normal/control samples.
#' @param auc_vector A numeric vector generated by [compute_AUC].
#' @param reference_table A data.frame reporting the location of CpG sites
#' (chromosome and genomic_position).
#' @param length_cutoff An integer to remove streches of differential
#' methylation shorter than cutoff.
#' @param pt_start Transition probability of the HSLM. Default is 0.05.
#' @param normdist Distance normalization parameter of the HSLM. Default is 1e5.
#' @param ratiosd Fraction between the standard deviation of AUC values of
#' differentially methylated sites and the total standard deviation of AUC
#' values. Default is 0.4.
#'
#' @importFrom stats mad median p.adjust sd wilcox.test
#' @export
whole_genome_segmentator <- function(tumor_table, control_table, auc_vector,
  reference_table, length_cutoff = 5, pt_start = 0.05, normdist = 1e5,
  ratiosd = 0.4){
  # check parameters
  if (length(ratiosd) != 1 | ratiosd <= 0 | ratiosd >= 1 ) {
    stop("ratiosd must be between 0 and 1")
  }
  if (length(normdist) != 1 | normdist < 1) {
    stop("normdist must be an integer between 1 and Inf. Suggested value is 1e5.")
  }
  if (length(pt_start) != 1 | pt_start <= 0 | pt_start >= 1 ) {
    stop("pt_start must be between 0 and 1")
  }
  tumor_table <- as.matrix(tumor_table)
  control_table <- as.matrix(control_table)
  diff_range_t <- diff(range(tumor_table, na.rm = TRUE))
  diff_range_c <- diff(range(control_table, na.rm = TRUE))
  if (diff_range_t <= 1 || diff_range_t > 100 ||
      diff_range_c <= 1 || diff_range_c > 100) {
    stop(paste0("For computation efficiency please convert tumor and control",
        "tables to percentage value."))
  } else {
    tumor_table <- round(tumor_table)
    control_table <- round(control_table)
    storage.mode(tumor_table) <- "integer"
    storage.mode(control_table) <- "integer"
  }
  assertthat::assert_that(is.data.frame(reference_table))
  assertthat::assert_that(length(reference_table) >= 2)
  if (nrow(tumor_table) != nrow(control_table) ||
      nrow(tumor_table) != nrow(reference_table)) {
    stop("Tumor and control tables must have as many rows as reference_table.")
  }

  # remove NA rows
  idx_not_NA <- which(!is.na(auc_vector))
  tumor_table <- tumor_table[idx_not_NA,, drop = FALSE]
  control_table <- control_table[idx_not_NA,, drop = FALSE]
  auc_vector <- auc_vector[idx_not_NA]
  reference_table <- reference_table[idx_not_NA,, drop = FALSE]

  # sort data
  idx <- order(reference_table[[1]], reference_table[[2]])
  reference_table <- reference_table[idx, ]
  tumor_table <- tumor_table[idx, ]
  control_table <- control_table[idx, ]

  tumor_beta_mean <- apply(tumor_table, 1, mean, na.rm = TRUE)
  control_beta_mean <- apply(control_table, 1, mean, na.rm = TRUE)
  auc_sd <- sd(auc_vector, na.rm = TRUE)

  chromosomes <- as.character(unique(reference_table[[1]]))
  all_chr_segs <- vector("list", length(chromosomes))
  for (i in seq_along(chromosomes)) {
    chr <- chromosomes[i]
    message(paste("Processing chromosome", chr))
    idx_chr <- which(reference_table[[1]] == chr)

    # 1) compute methylation states (1,2,3)
    meth_states <- meth_state_finder(auc_vector[idx_chr],
      reference_table[[2]][idx_chr], auc_sd, length_cutoff, pt_start, normdist,
      ratiosd)

    # 2) find segments
    single_chr_segs <- segmentator(meth_states, tumor_beta_mean[idx_chr],
      control_beta_mean[idx_chr])

    meth_states <- with(single_chr_segs, rep(state, nseg))
    rle_states <- S4Vectors::Rle(meth_states)
    rle_start <- S4Vectors::start(rle_states)
    rle_end <- S4Vectors::end(rle_states)

    all_chr_segs[[i]] <- data.frame(stringsAsFactors = FALSE,
      chr = reference_table[[1]][idx_chr[rle_start]],
      start = reference_table[[2]][idx_chr[rle_start]],
      end = reference_table[[2]][idx_chr[rle_end]],
      single_chr_segs)
  }
  all_chr_segs <- dplyr::bind_rows(all_chr_segs)
  all_chr_segs$q_value <- p.adjust(all_chr_segs$p_value, "fdr")
  return(all_chr_segs)
}

#' Define Differentially Methylated Regions
#'
#' Given a set of methylation states produced by [meth_states_finder] and a set of
#' methylation (beta) values differences between tumor and control samples, this fucntion
#' defines stretches of equal states (segments).
#'
#' @param meth_states a vector of methylation states (1, 2, 3, NA)
#' @param tumor_beta_mean a vector of average methylation (beta) values
#' @param control_beta_mean a vector of average methylation (beta) values
#'
#' @return a data.frame with information about segments
#' @keywords internal
segmentator <- function(meth_states, tumor_beta_mean, control_beta_mean){
  assertthat::assert_that(length(tumor_beta_mean) == length(control_beta_mean))
  beta_diff <- tumor_beta_mean - control_beta_mean

  rle_states <- S4Vectors::Rle(meth_states)
  rle_length <- S4Vectors::runLength(rle_states)
  rle_value  <- S4Vectors::runValue(rle_states)
  rle_n      <- S4Vectors::nrun(rle_states)
  rle_start  <- S4Vectors::start(rle_states)
  rle_end    <- S4Vectors::end(rle_states)

  values <- matrix(NA, rle_n, 2,
    dimnames = list(NULL, c("avg_beta_diff", "p_value")))

  for (i in seq_len(rle_n)) {
    values[i, 1] <- mean(beta_diff[rle_start[i]: rle_end[i]], na.rm = TRUE)

    if (rle_value[i] != 2) {
      hypothesis <- dplyr::if_else(rle_value[i] == 3, "greater", "less")
      values[i, 2] <- suppressWarnings(tryCatch(wilcox.test(
        tumor_beta_mean[rle_start[i]:rle_end[i]],
        control_beta_mean[rle_start[i]:rle_end[i]], hypothesis)$p.value,
        error = function(e) NA))
    }
  }
  return(data.frame(nseg = rle_length, state = rle_value, values))
}
