#' Find Differentially Methylated Regions across whole genome
#'
#' A function to define DMR for an entire genome
#'
#' @param tumor_table A matrix of beta-values from tumor samples.
#' @param control_table A matrix of beta-values from normal/control samples.
#' @param auc_vector An numeric vector generated by [compute_AUC].
#' @param reference_table A data.frame reporting location of CpG probes (chromosome
#' and genomic_position)
#' @param length_cutoff An integer to remove streches of differential
#' methylation shorter than cutoff
#'
#' @importFrom stats mad median p.adjust sd wilcox.test
#' @export
whole_genome_segmentator <- function(tumor_table, control_table, auc_vector,
  reference_table, length_cutoff = 5){
  # check parameters
  tumor_table <- as.matrix(tumor_table)
  control_table <- as.matrix(control_table)
  stopifnot(is.numeric(auc_vector))
  stopifnot(is.data.frame(reference_table))
  if (nrow(tumor_table) != nrow(reference_table) ||
    nrow(control_table) != nrow(reference_table)) {
    stop(paste0("tumor_table and control_table must have same number",
        "of rows as reference_table"))
  }
  tumor_is_fraction <- all(tumor_table >= 0, na.rm = TRUE) &&
    all(tumor_table <= 1, na.rm = TRUE)
  control_is_fraction <- all(control_table >= 0, na.rm = TRUE) &&
    all(control_table <= 1, na.rm = TRUE)
  if (tumor_is_fraction && control_is_fraction) {
    tumor_table <- round(tumor_table*100)
    control_table <- round(control_table*100)
    storage.mode(control_table) <- "integer"
    storage.mode(tumor_table) <- "integer"
  } else {
    stop("tumor_table and control_table must have fraction values")
  }

  # remove NA rows
  idx_not_NA <- which(!is.na(auc_vector))
  tumor_table <- tumor_table[idx_not_NA,,drop = FALSE]
  control_table <- control_table[idx_not_NA,,drop = FALSE]
  auc_vector <- auc_vector[idx_not_NA]
  reference_table <- reference_table[idx_not_NA,,drop = FALSE]

  # sort data
  idx <- order(reference_table[[1]], reference_table[[2]])
  reference_table <- reference_table[idx, ]
  tumor_table <- tumor_table[idx, ]
  control_table <- control_table[idx, ]

  tumor_beta_mean <- apply(tumor_table, 1, mean, na.rm = TRUE)
  control_beta_mean <- apply(control_table, 1, mean, na.rm = TRUE)
  auc_sd <- sd(auc_vector, na.rm=TRUE)
  pt_start <- 0.05

  chromosomes <- unique(reference_table[[1]])
  all_chr_segs <- do.call("rbind", lapply(chromosomes, function(chr) {
    idx_chr <- which(reference_table[[1]] == chr)

    # 1) compute methylation states (1,2,3)
    meth_states <- meth_state_finder(auc_vector[idx_chr],
      reference_table[[2]][idx_chr], auc_sd, pt_start, length_cutoff)

    # 2) find segments for each chromosome
    single_chr_segs <- segmentator(meth_states, tumor_beta_mean[idx_chr],
      control_beta_mean[idx_chr])

    meth_states <- with(single_chr_segs, rep(state, nseg))
    rle_states <- S4Vectors::Rle(meth_states)
    rle_start <- S4Vectors::start(rle_states)
    rle_end <- S4Vectors::end(rle_states)

    return(data.frame(
      chr = reference_table[[1]][idx_chr[rle_start]],
      start = reference_table[[2]][idx_chr[rle_start]],
      end = reference_table[[2]][idx_chr[rle_end]],
      single_chr_segs))
  }))
  all_chr_segs$q_value <- p.adjust(all_chr_segs$p_value, "fdr")
  return(all_chr_segs)
}

#' Define Differentially Methylated Regions
#'
#' Given a set of methylation states produced by [meth_states_finder] and a set of
#' methylation (beta) values differences between tumor and control samples, this fucntion
#' defines stretches of equal states (segments).
#'
#' @param meth_states a vector of methylation states (1, 2, 3, NA)
#' @param tumor_beta_mean a vector of average methylation (beta) values
#' @param control_beta_mean a vector of average methylation (beta) values
#'
#' @return a data.frame with information about segments
#' @keywords internal
segmentator <- function(meth_states, tumor_beta_mean, control_beta_mean){
  stopifnot(all(meth_states %in% c(NA, 1:3)))
  stopifnot(is.numeric(tumor_beta_mean))
  stopifnot(is.numeric(control_beta_mean))
  stopifnot(length(tumor_beta_mean) == length(control_beta_mean))
  beta_diff <- tumor_beta_mean - control_beta_mean

  rle_states <- S4Vectors::Rle(meth_states)
  rle_length <- S4Vectors::runLength(rle_states)
  rle_value  <- S4Vectors::runValue(rle_states)
  rle_n      <- S4Vectors::nrun(rle_states)
  rle_start  <- S4Vectors::start(rle_states)
  rle_end    <- S4Vectors::end(rle_states)

  values <- matrix(nrow = rle_n, ncol = 2, dimnames = list(NULL, c("avg_beta_diff", "p_value")))
  for (i in seq_len(rle_n)) {
    values[i, 1] <- mean(beta_diff[rle_start[i]: rle_end[i]], na.rm = TRUE)
    if (!(rle_value[i] == 2 | is.na(rle_value[i]))) {
      hypothesis <- ifelse(rle_value[i] == 3, "greater", "less")
      values[i, 2] <- suppressWarnings(tryCatch(wilcox.test(
            tumor_beta_mean[rle_start[i]:rle_end[i]],
            control_beta_mean[rle_start[i]:rle_end[i]], hypothesis)$p.value,
          error = function(e) NA))
    }
  }
  results <- data.frame(nseg = rle_length, state = rle_value, values)
  return(results)
}
